govcd/cse.go

```diff
@@ -0,0 +1,379 @@
+	}
+
+	cseSubcomponents, err := getCseComponentsVersions(clusterSettings.CseVersion)
+	if err != nil {
+		return "", err
+	}
+
+	internalSettings, err := clusterSettings.toCseClusterSettingsInternal(*org)
+	if err != nil {
+		return "", fmt.Errorf("error creating the CSE Kubernetes cluster: %s", err)
+	}
+
+	payload, err := internalSettings.getUnmarshaledRdePayload()
+	if err != nil {
+		return "", err
+	}
+
+	rde, err := createRdeAndGetFromTask(org.client, "vmware", "capvcdCluster", cseSubcomponents.CapvcdRdeTypeVersion,
+		types.DefinedEntity{
+			EntityType: internalSettings.RdeType.ID,
+			Name:       internalSettings.Name,
+			Entity:     payload,
+		}, &TenantContext{
+			OrgId:   org.Org.ID,
+			OrgName: org.Org.Name,
+		})
+	if err != nil {
+		return "", fmt.Errorf("error creating the CSE Kubernetes cluster: %s", err)
+	}
+
+	return rde.DefinedEntity.ID, nil
+}
+
+// CseGetKubernetesClusterById retrieves a CSE Kubernetes cluster from VCD by its unique ID
+func (vcdClient *VCDClient) CseGetKubernetesClusterById(id string) (*CseKubernetesCluster, error) {
+	return getCseKubernetesClusterById(&vcdClient.Client, id)
+}
+
+// CseGetKubernetesClustersByName retrieves all the CSE Kubernetes clusters from VCD with the given name that belong to the receiver Organization.
+// Note: The clusters retrieved won't have a valid ETag to perform operations on them. Use VCDClient.CseGetKubernetesClusterById for that instead.
+func (org *Org) CseGetKubernetesClustersByName(cseVersion semver.Version, name string) ([]*CseKubernetesCluster, error) {
+	cseSubcomponents, err := getCseComponentsVersions(cseVersion)
+	if err != nil {
+		return nil, err
+	}
+
+	rdes, err := getRdesByName(org.client, "vmware", "capvcdCluster", cseSubcomponents.CapvcdRdeTypeVersion, name)
+	if err != nil {
+		return nil, err
+	}
+	var clusters []*CseKubernetesCluster
+	for _, rde := range rdes {
+		if rde.DefinedEntity.Org != nil && rde.DefinedEntity.Org.ID == org.Org.ID {
+			cluster, err := cseConvertToCseKubernetesClusterType(rde)
+			if err != nil {
+				return nil, err
+			}
+			clusters = append(clusters, cluster)
+		}
+	}
+	return clusters, nil
+}
+
+// getCseKubernetesClusterById retrieves a CSE Kubernetes cluster from VCD by its unique ID
+func getCseKubernetesClusterById(client *Client, clusterId string) (*CseKubernetesCluster, error) {
+	rde, err := getRdeById(client, clusterId)
+	if err != nil {
+		return nil, err
+	}
+	return cseConvertToCseKubernetesClusterType(rde)
+}
+
+// Refresh gets the latest information about the receiver CSE Kubernetes cluster and updates its properties.
+// All cached fields such as the supported OVAs list (from CseKubernetesCluster.GetSupportedUpgrades) are also cleared.
+func (cluster *CseKubernetesCluster) Refresh() error {
+	refreshed, err := getCseKubernetesClusterById(cluster.client, cluster.ID)
+	if err != nil {
+		return fmt.Errorf("failed refreshing the CSE Kubernetes Cluster: %s", err)
+	}
+	*cluster = *refreshed
```

Comment:
Any particular reason for this notation? IIUC, `cluster = refreshed` should work fine as well

Reply:
To be honest I thought the same but the IDE warned me it wouldn't work. Doing a bit of research the reason looks like this:

- Doing `cluster = refreshed` overrides the "cluster" address with the "refreshed" address, but "cluster" is scoped to the method, so outside of it does nothing, the address goes back to the previous value. I think the reason is, "cluster" is a copy of the address, so changing it will do nothing.

- On the other hand, doing `*cluster = *refreshed` overrides the "cluster" pointed contents with the "cluster" pointed contents, so as the address remains the same (due to reason above), it works.

[Here](https://stackoverflow.com/questions/38105543/reassigning-in-pointer-method-receiver) probably it's explained a bit better than I did 😅